yellow-paper/docs/circuits/private-kernel-inner.md

```diff
@@ -0,0 +1,282 @@
+   - Public call requests.
+
+   > Note that _counter_start_ is used in the above steps for public call requests to ensure their correct ordering. At this point, the _counter_end_ of public call request is unknown. Both counters will be [recalibrated](./private-kernel-tail.md#recalibrating-counters) in the tail circuit following the simulation of all public function calls.
+
+### Validating Public Inputs
+
+#### Verifying the accumulated data.
+
+It checks that the hashes and the lengths for both encrypted and unencrypted logs are accumulated as follows:
+
+- New log hash = `hash(prev_hash, cur_hash)`
+  - If either hash is zero, the new hash will be `prev_hash | cur_hash`
+- New log length = `prev_length + cur_length`
+
+#### Verifying the transient accumulated data.
+
+1. It verifies that the following values match the result of combining the values in the previous iteration's public inputs with those in the private call data:
+
+   - Note hash contexts.
+   - Nullifier contexts.
+   - L2-to-L1 message contexts.
+   - New contract contexts.
+   - Read requests.
+   - Public call requests.
+
+2. For the newly added note hashes from private function circuits' public inputs, this circuit also checks that each is associated with a nullifier counter, provided as a hint via the private inputs. The nullifier counter can be:
+
+   - Zero: if the note is not nullified in the same transaction.
+   - Greater than zero: if the note is nullified in the same transaction.
+     - This value must be greater than the counter of the note hash.
+
+   > Nullifier counters are used in the [reset private kernel circuit](./private-kernel-reset.md) to ensure a read happens **before** a transient note is nullified.
+
+   > Zero can be used to indicate a non-existing transient nullifier, as this value can never serve as the counter of a nullifier. It corresponds to the _counter_start_ of the first function call.
+
+3. It verifies that the private call requests include:
+
+   - All requests from the previous iteration's public inputs excluding the top one.
+   - All requests present in the private call data, appended to the above in **reverse** order.
+
+   > Ensuring the chronological execution of call requests is vital, requiring them to be arranged in reverse order. This becomes particularly crucial when calling a contract deployed earlier within the same transaction.
+
+#### Verifying the constant data.
+
+It verifies that the constant data matches the one in the previous iteration's public inputs.
+
+## Private Inputs
+
+### Previous Kernel
+
+The data of the previous kernel iteration:
+
+- Proof of the kernel circuit. It must be one of the following:
+  - [Initial private kernel circuit](./private-kernel-initial.md).
+  - Inner private kernel circuit.
+  - [Reset private kernel circuit](./private-kernel-reset.md).
+- Public inputs of the proof.
+- Verification key of the kernel circuit.
+- Membership witness for the verification key.
+
+### Private Call Data
```

Comment:
If this information is the same as the section in `private-kernel-initial.md`, we should define it only *once* and link to it in additional sections.

Redundant definitions of the same data structures can lead to errors as both definitions need to be kept in sync with one another